src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -561,16 +567,32 @@ void ObjectMonitor::enter_with_contention_mark(JavaThread *current, ObjectMonito
 
 ObjectMonitor::TryLockResult ObjectMonitor::TryLock(JavaThread* current) {
   void* own = owner_raw();
-  if (own != nullptr) return TryLockResult::HasOwner;
-  if (try_set_owner_from(nullptr, current) == nullptr) {
-    assert(_recursions == 0, "invariant");
-    return TryLockResult::Success;
+  void* first_own = own;
+
+  for (;;) {
+    if (own == DEFLATER_MARKER) {
+      if (TryLockI(current)) {
+        assert(_recursions == 0, "invariant");
+        return TryLockResult::Success;
+      } else {
+        // Deflation won or change of owner; dont spin
+        break;
+      }
+    } else if (own == nullptr) {
+      void* prev_own = try_set_owner_from(nullptr, current);
+      if (prev_own == nullptr) {
+        assert(_recursions == 0, "invariant");
+        return TryLockResult::Success;
+      } else {
+        // The lock had been free momentarily, but we lost the race to the lock.
+        own = prev_own;
```

Comment:
It could be the deflator (or someone else). Anyhow, we will retry.

Reply:
Now that I'm reading the code more slowly, I see that the reason for this loop is the deflator.  Can you add this to this comment, so I can read this faster next time :)

     // The lock had been free momentarily, but we lost the race to the lock. Retry in case the lock was acquired by the deflator.